rust/otap-dataflow/crates/otlp/src/otlp_batch_processor.rs

```diff
@@ -24,57 +24,51 @@ pub trait HierarchicalBatchSplit: Sized {
 impl HierarchicalBatchSplit for ExportTraceServiceRequest {
     fn split_into_batches(mut self, max_batch_size: usize) -> Vec<Self> {
         let mut batches = Vec::new();
-
-        // Accumulates resource groups and their contents until the batch reaches max_batch_size leaf items.
-        // When full, this batch is pushed to the output and a new one is started.
         let mut current_batch = ExportTraceServiceRequest {
             resource_spans: Vec::new(),
         };
         let mut current_span_count = 0;
 
-        // ToDo: The current implementation is recreating the entire hierarchy of ResourceSpans, ScopeSpans, Spans but we should probably avoid most of those allocations.
         for mut rs in self.resource_spans.drain(..) {
             let mut res = ResourceSpans {
                 resource: rs.resource.take(),
                 scope_spans: Vec::new(),
                 schema_url: rs.schema_url.clone(),
             };
             for mut ss in rs.scope_spans.drain(..) {
-                let mut sc = ScopeSpans {
-                    scope: ss.scope.take(),
-                    spans: Vec::new(),
-                    schema_url: ss.schema_url.clone(),
-                };
                 while !ss.spans.is_empty() {
-                    // Number of items that can still be added to the current batch before reaching max_batch_size
                     let remaining = max_batch_size - current_span_count;
-                    let take = remaining.min(ss.spans.len());
-                    sc.spans.extend(ss.spans.drain(..take));
-                    current_span_count += take;
-
-                    if !sc.spans.is_empty() {
-                        res.scope_spans.push(sc.clone());
-                        sc.spans.clear();
-                    }
+                    let take_amount = remaining.min(ss.spans.len());
+                    let mut this_spans = ss.spans.split_off(ss.spans.len() - take_amount);
+                    // The original `ss.spans` now only has the left-overs
+
+                    let mut this_scope = ScopeSpans {
+                        scope: ss.scope.clone(),
+                        spans: Vec::new(),
+                        schema_url: ss.schema_url.clone(),
+                    };
+                    this_scope.spans.append(&mut this_spans);
```

Comment:
You can reuse the vec allocated by `split_off` instead of throwing it away by moving its contents to `this_scope.spans`.

```suggestion
                    std::mem::swap(&mut this_scope.spans, &mut this_spans);
```